internal/xds/translator/cluster.go

```diff
@@ -315,13 +312,38 @@ func buildXdsClusterLoadAssignment(clusterName string, destSettings []*ir.Destin
 	return &endpointv3.ClusterLoadAssignment{ClusterName: clusterName, Endpoints: localities}
 }
 
-func buildTypedExtensionProtocolOptions() map[string]*anypb.Any {
-	protocolOptions := httpv3.HttpProtocolOptions{
-		UpstreamProtocolOptions: &httpv3.HttpProtocolOptions_ExplicitHttpConfig_{
+func buildTypedExtensionProtocolOptions(args *xdsClusterArgs) map[string]*anypb.Any {
+	isHTTP2 := false
+	for _, ds := range args.settings {
+		if ds.Protocol == ir.GRPC ||
+			ds.Protocol == ir.HTTP2 {
+			isHTTP2 = true
+			break
+		}
+	}
+
+	hasCommonHTTPOptions := args.timeout != nil && args.timeout.HTTP != nil &&
+		(args.timeout.HTTP.MaxConnectionDuration != nil || args.timeout.HTTP.ConnectionIdleTimeout != nil)
+
+	if !hasCommonHTTPOptions && !isHTTP2 {
+		return nil
+	}
+
+	protocolOptions := httpv3.HttpProtocolOptions{}
+
+	if hasCommonHTTPOptions {
+		protocolOptions.CommonHttpProtocolOptions = &corev3.HttpProtocolOptions{
+			IdleTimeout:           durationpb.New(args.timeout.HTTP.ConnectionIdleTimeout.Duration),
+			MaxConnectionDuration: durationpb.New(args.timeout.HTTP.MaxConnectionDuration.Duration),
```

Comment:
these two duration should be checked (whether is nil ptr) before assign values to`IdleTimeout` and `MaxConnectionDuration`, nor it will raise a nil pointer panic.

```yaml
http:
- name: "first-listener"
  #...
  routes:
  - name: "first-route"
    backendTimeout:
      tcp:
        connectTimeout: "31s"
      http:
        connectionIdleTimeout: "32s"
        #maxConnectionDuration: "33s" # paniced
    #...
```